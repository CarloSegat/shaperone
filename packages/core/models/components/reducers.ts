import produce from 'immer'
import { NamedNode } from 'rdf-js'
import type { ComponentsState, ComponentState, RenderFunc, Decorator, Component } from './index'

type _Component = Omit<ComponentState, 'loading' | 'loadingFailed'>

export default {
  loading(components: ComponentsState, editor: NamedNode): ComponentsState {
    return produce(components, (draft) => {
      draft[editor.value].loading = true
    })
  },
  loadingFailed(components: ComponentsState, { editor, reason }: { editor: NamedNode; reason: string }): ComponentsState {
    return produce(components, (draft) => {
      draft[editor.value].loading = false
      draft[editor.value].loadingFailed = {
        reason,
      }
    })
  },
  loaded(components: ComponentsState, { editor, render } : {editor: NamedNode; render: RenderFunc<any, any, any>}): ComponentsState {
    return produce(components, (draft) => {
      draft[editor.value].loading = false
      draft[editor.value].render = render
      draft[editor.value].loadingFailed = undefined
    })
  },
  removeComponents(components: ComponentsState, toRemove: NamedNode[]) {
    return produce(components, (newComponents) => {
      for (const editor of toRemove) {
        delete newComponents[editor.value]
      }
    })
  },
  pushComponents(components: ComponentsState, toAdd: Record<string, _Component> | _Component[]): ComponentsState {
    return produce(components, (newComponents) => {
      const addedArray = Array.isArray(toAdd) ? toAdd : Object.values(toAdd)

      for (const component of addedArray) {
        const previous = components[component.editor.value]
        const shouldAddComponent = !previous ||
        (previous.lazyRender && component.lazyRender && previous.lazyRender !== component.lazyRender) ||
        (previous.render && component.render && previous.render !== component.render) ||
        ((previous.render && !component.render) || (previous.lazyRender || !component.lazyRender))

        if (shouldAddComponent) {
          newComponents[component.editor.value] = {
            ...component,
            loading: false,
          }
        }
      }
    })
  },
  decorate<T extends Component = Component>(components: ComponentsState, decorator: Decorator<T>) {
    return produce(components, (decorated) => {
      for (const [key, component] of Object.entries(components)) {
        if (decorator.applicableTo(component)) {
          decorated[key] = {
            ...component,
            ...decorator.decorate(component as unknown as T),
          }
        }
      }
    })
  },
}
