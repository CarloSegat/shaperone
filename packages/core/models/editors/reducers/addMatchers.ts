import produce from 'immer'
import { dash } from '@tpluscode/rdf-ns-builders'
import type { EditorsState, MultiEditor, SingleEditor } from '../index'
import { EditorMeta } from '../lib/EditorMeta'
import { combine } from '../lib/decorators'

type AnyEditor = SingleEditor<any> | MultiEditor

export function addMatchers(state: EditorsState, editors: Record<string, AnyEditor> | AnyEditor[]): EditorsState {
  return produce(state, (draft) => {
    const editorsArray = Array.isArray(editors) ? editors : Object.values(editors)

    for (const editor of editorsArray) {
      const { term } = editor
      const key = editor.term.value
      const meta = new EditorMeta(state.metadata.node(term))
      const entry = {
        ...editor as any,
        meta,
      }

      draft.allEditors[key] = entry

      if (meta.types.has(dash.MultiEditor)) {
        draft.multiEditors[key] = combine(state.decorators, entry)
      } else {
        draft.singleEditors[key] = combine(state.decorators, entry)
      }
    }
  })
}
