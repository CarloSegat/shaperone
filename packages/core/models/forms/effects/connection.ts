import type { Store } from '../../../state'
import { BaseParams } from '../../index'

export function connect(store: Store) {
  const dispatch = store.getDispatch()
  return ({ form }: BaseParams) => {
    const { forms, editors, shapes, resources } = store.getState()
    const shapesState = shapes.get(form)
    const formState = forms.get(form)

    const rootPointer = resources.get(form)?.rootPointer

    if (rootPointer && formState) {
      dispatch.forms.createFocusNodeState({
        form,
        focusNode: rootPointer,
        editors,
        shouldEnableEditorChoice: formState.shouldEnableEditorChoice,
        shapes: shapesState?.shapes || [],
        shape: shapesState?.preferredRootShape,
        replaceStack: true,
      })
    }

    dispatch.resources.connect(form)
    dispatch.shapes.connect(form)
  }
}

export function disconnect(store: Store) {
  const dispatch = store.getDispatch()
  return (form: symbol) => {
    dispatch.resources.disconnect(form)
    dispatch.shapes.disconnect(form)
  }
}
