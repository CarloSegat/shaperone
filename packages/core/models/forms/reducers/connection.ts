import type { State } from '../index'

export function connect(map: State, form: symbol): State {
  map.instances.set(form, {
    focusNodes: {},
    focusStack: [],
    shouldEnableEditorChoice: () => true,
  })
  return map
}

export function disconnect(map: State, form: symbol): State {
  map.instances.delete(form)
  return map
}
