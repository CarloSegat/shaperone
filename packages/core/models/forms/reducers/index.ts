import type { FormState, State } from '../index'
import type { MultiEditor, SingleEditor, Editor } from '../../editors'

export interface BaseParams {
  form: symbol
}

export interface SharedParams {
  state: FormState
  editors: Editor<SingleEditor>[]
  multiEditors: Editor<MultiEditor>[]
}

export function formStateReducer<T>(reduceForm: (sharedParams: SharedParams, payload: T) => FormState) {
  return (state: State, params: T & BaseParams): State => {
    const formState = state.instances.get(params.form)
    if (formState) {
      state.instances.set(params.form, reduceForm({
        state: formState,
        editors: state.singleEditors,
        multiEditors: state.multiEditors,
      }, params))
    }

    return state
  }
}
