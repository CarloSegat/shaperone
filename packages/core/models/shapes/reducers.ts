import { DatasetCore } from 'rdf-js'
import cf, { AnyPointer } from 'clownface'
import produce from 'immer'
import { rdf, sh } from '@tpluscode/rdf-ns-builders'
import RdfResource from '@tpluscode/rdfine/RdfResource'
import { Shape, ShapeMixin } from '@rdfine/shacl'
import { BaseParams, formStateReducer } from '../index'
import type { ShapeState } from '.'

export interface SetShapesGraphParams extends BaseParams {
  shapesGraph: DatasetCore | AnyPointer
}

export const setGraph = formStateReducer((state: ShapeState, { shapesGraph }: SetShapesGraphParams) => produce(state, (draft) => {
  if (state.shapesGraph === shapesGraph) {
    return
  }

  const shapesPointer = 'match' in shapesGraph ? cf({ dataset: shapesGraph }) : shapesGraph.any()
  const shapes = shapesPointer
    .has(rdf.type, [sh.Shape, sh.NodeShape])
    .map(pointer => RdfResource.factory.createEntity<Shape>(pointer, [ShapeMixin]))

  draft.shapesGraph = shapesPointer
  draft.shapes = shapes
}))
