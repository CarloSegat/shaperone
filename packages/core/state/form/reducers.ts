import type { FocusNodeState, FormState, PropertyObjectState } from '../../state'
import type { PropertyGroup, PropertyShape, Shape } from '@rdfine/shacl'
import { FocusNode } from '../../index'
import { initialiseFocusNode } from '../../lib/stateBuilder'
import { sh } from '@tpluscode/rdf-ns-builders'
import { NamedNode, Term } from 'rdf-js'
import { blankNode, literal } from '@rdf-esm/data-model'
import RdfResource from '@tpluscode/rdfine/RdfResource'

interface BaseParams {
  focusNode: FocusNode
  property: PropertyShape
}

interface SelectEditorParams extends BaseParams {
  value: Term
  editor: NamedNode
}

export function selectEditor(state: FormState, { focusNode, property, value, editor }: SelectEditorParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]
  const properties = focusNodeState.properties.map(prop => {
    if (!prop.shape.id.equals(property.id)) {
      return prop
    }

    const objects = prop.objects.map(o => {
      if (o.object.term.equals(value)) {
        return {
          ...o,
          selectedEditor: editor,
        }
      }

      return o
    })

    return {
      ...prop,
      objects,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

interface UpdateObjectParams extends BaseParams {
  oldValue: Term
  newValue: Term
}

export function updateObject(state: FormState, { focusNode, property, oldValue, newValue }: UpdateObjectParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]
  const properties = focusNodeState.properties.map(prop => {
    if (!prop.shape.id.equals(property.id)) {
      return prop
    }

    const objects = prop.objects.map((o): PropertyObjectState => {
      if (o.object.term.equals(oldValue)) {
        return {
          ...o,
          object: focusNodeState.focusNode.node(newValue),
        }
      }

      return o
    })

    focusNodeState.focusNode
      .deleteOut(property.path.id)
      .addOut(property.path.id, objects.map(o => o.object))

    return {
      ...prop,
      objects,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

function defaultValueNode(property: PropertyShape): Term {
  if (property.get(sh.class) || property.get(sh.nodeKind)?.id.equals(sh.IRI)) {
    return blankNode()
  }

  return literal('')
}

export function addObject(state: FormState, { focusNode, property }: BaseParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]

  const object = property.defaultValue ? focusNodeState.focusNode.node(property.defaultValue) : focusNodeState.focusNode.node(defaultValueNode(property))

  focusNodeState.focusNode.addOut(property.path.id, object)

  const properties = focusNodeState.properties.map(currentProperty => {
    if (!currentProperty.shape.id.equals(property.id)) {
      return currentProperty
    }
    if (currentProperty.objects.find(o => o.object.term.equals(object.term))) {
      return currentProperty
    }

    const editors = [...state.editorMap.values()]
      .map(({ match }) => match(property, object))
      .filter(match => match.score === null || match.score > 0)
      .sort((left, right) => left.score! - right.score!)

    const maxReached = (property.getNumber(sh.maxCount) || Number.POSITIVE_INFINITY) <= currentProperty.objects.length + 1
    const newObject: PropertyObjectState = {
      object,
      editors,
      selectedEditor: editors[0]?.editor,
    }

    return {
      ...currentProperty,
      objects: [
        ...currentProperty.objects,
        newObject,
      ],
      maxReached,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

interface RemoveObjectParams extends BaseParams {
  object: PropertyObjectState
}

export function removeObject(state: FormState, { focusNode, property, object }: RemoveObjectParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]

  const properties = focusNodeState.properties.map(currentProperty => {
    if (!currentProperty.shape.id.equals(property.id)) {
      return currentProperty
    }

    const objects = currentProperty.objects.filter(o => !o.object.term.equals(object.object.term))

    focusNodeState.focusNode
      .deleteOut(property.path.id)
      .addOut(property.path.id, objects.map(o => o.object))

    const maxReached = (property.getNumber(sh.maxCount) || Number.POSITIVE_INFINITY) <= objects.length

    return {
      ...currentProperty,
      objects,
      maxReached,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

export function editorLoading(state: FormState, { editor }: { editor: NamedNode }): FormState {
  return {
    ...state,
    editors: {
      ...state.editors,
      [editor.value]: {
        loaded: false,
      },
    },
  }
}

export function editorLoaded(state: FormState, { editor }: { editor: NamedNode }): FormState {
  return {
    ...state,
    editors: {
      ...state.editors,
      [editor.value]: {
        loaded: true,
      },
    },
  }
}

export function resetEditors(state: FormState): FormState {
  return {
    ...state,
    editors: {},
  }
}

export function pushFocusNode(state: FormState, { focusNode, property }: { focusNode: FocusNode; property: PropertyShape }): FormState {
  const propertyTargetClass = property.get(sh.class)
  if (!propertyTargetClass) {
    return state
  }

  const shapePointer = state.rootShape!._selfGraph.has(sh.targetClass, propertyTargetClass.id).toArray()
  if (!shapePointer.length) {
    return state
  }

  const shape = RdfResource.factory.createEntity<Shape>(shapePointer[0])

  return {
    ...state,
    focusStack: [...state.focusStack, focusNode],
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: initialiseFocusNode({ focusNode, shape, state }),
    },
  }
}

export function truncateFocusNodes(state: FormState, { focusNode }: { focusNode: FocusNode }): FormState {
  const topNodeIndex = state.focusStack.findIndex(fn => fn.term.equals(focusNode.term))
  if (topNodeIndex < 0) {
    return state
  }

  return {
    ...state,
    focusStack: state.focusStack.slice(0, topNodeIndex),
  }
}

export function popFocusNode(state: FormState): FormState {
  return {
    ...state,
    focusStack: state.focusStack.slice(0, -1),
  }
}

export function selectGroup(state: FormState, { group, focusNode }: { focusNode: FocusNode; group?: PropertyGroup }): FormState {
  const groups = state.focusNodes[focusNode.value].groups.map(g => {
    let selected = false
    if (!group && !g.group) {
      selected = true
    } else {
      selected = group?.id.equals(g.group?.id) || false
    }

    return {
      ...g,
      selected,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...state.focusNodes[focusNode.value],
        groups,
      },
    },
  }
}

export function initialize(state: FormState, params: { focusNode: FocusNode; rootShape?: Shape }): FormState {
  const shape = state.rootShape || params.rootShape

  const { focusNode } = params

  return {
    ...state,
    rootShape: shape,
    focusStack: [focusNode],
    focusNodes: {
      [focusNode.value]: initialiseFocusNode({
        shape,
        state,
        focusNode,
      }),
    },
  }
}

export function recalculateFocusNodes(state: FormState, shape: Shape): FormState {
  const focusNodes = Object.values(state.focusNodes).reduce<Record<string, FocusNodeState>>((obj, focusNode) => {
    const selectedGroup = focusNode.groups.find(g => g.selected)?.group?.id.value

    return {
      ...obj,
      [focusNode.focusNode.value]: initialiseFocusNode({
        shape,
        state,
        focusNode: focusNode.focusNode,
        selectedGroup,
      }),
    }
  }, {})

  return {
    ...state,
    rootShape: shape,
    focusNodes,
  }
}
