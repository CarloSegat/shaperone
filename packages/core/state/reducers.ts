import type { FormState, PropertyObjectState } from '../state'
import type { PropertyShape, Shape } from '@rdfine/shacl'
import * as ns from '@tpluscode/rdf-ns-builders'
import { FocusNode } from '../index'
import { initialiseFocusNode } from '../lib/stateBuilder'
import { sh } from '@tpluscode/rdf-ns-builders'
import { NamedNode, Term } from 'rdf-js'

interface BaseParams {
  focusNode: FocusNode
  property: PropertyShape
}

interface SelectEditorParams extends BaseParams {
  value: Term
  editor: NamedNode
}

export function selectEditor(state: FormState, { focusNode, property, value, editor }: SelectEditorParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]
  const properties = focusNodeState.properties.map(prop => {
    if (!prop.shape.id.equals(property.id)) {
      return prop
    }

    const objects = prop.objects.map(o => {
      if (o.object.term.equals(value)) {
        return {
          ...o,
          selectedEditor: editor,
        }
      }

      return o
    })

    return {
      ...prop,
      objects,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

interface UpdateObjectParams extends BaseParams {
  oldValue: Term
  newValue: Term
}

export function updateObject(state: FormState, { focusNode, property, oldValue, newValue }: UpdateObjectParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]
  const properties = focusNodeState.properties.map(prop => {
    if (!prop.shape.id.equals(property.id)) {
      return prop
    }

    const objects = prop.objects.map((o): PropertyObjectState => {
      if (o.object.term.equals(oldValue)) {
        return {
          ...o,
          object: focusNodeState.focusNode.node(newValue),
        }
      }

      return o
    })

    focusNodeState.focusNode
      .deleteOut(property.path.id)
      .addOut(property.path.id, objects.map(o => o.object))

    return {
      ...prop,
      objects,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

export function addObject(state: FormState, { focusNode, property }: BaseParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]

  const object = property.defaultValue ? focusNodeState.focusNode.node(property.defaultValue) : focusNodeState.focusNode.literal('')

  focusNodeState.focusNode.addOut(property.path.id, object)

  const properties = focusNodeState.properties.map(currentProperty => {
    if (!currentProperty.shape.id.equals(property.id)) {
      return currentProperty
    }
    if (currentProperty.objects.find(o => o.object.term.equals(object.term))) {
      return currentProperty
    }

    const editors = [...state.editorMap.values()]
      .map(({ match }) => match(property, object))
      .filter(match => match.score === null || match.score > 0)
      .sort((left, right) => left.score! - right.score!)

    const maxReached = (property.getNumber(sh.maxCount) || Number.POSITIVE_INFINITY) <= currentProperty.objects.length + 1
    const newObject: PropertyObjectState = {
      object,
      editors,
      selectedEditor: editors[0]?.editor,
    }

    return {
      ...currentProperty,
      objects: [
        ...currentProperty.objects,
        newObject,
      ],
      maxReached,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

interface RemoveObjectParams extends BaseParams {
  object: PropertyObjectState
}

export function removeObject(state: FormState, { focusNode, property, object }: RemoveObjectParams): FormState {
  const focusNodeState = state.focusNodes[focusNode.value]

  const properties = focusNodeState.properties.map(currentProperty => {
    if (!currentProperty.shape.id.equals(property.id)) {
      return currentProperty
    }

    const objects = currentProperty.objects.filter(o => !o.object.term.equals(object.object.term))

    focusNodeState.focusNode
      .deleteOut(property.path.id)
      .addOut(property.path.id, objects.map(o => o.object))

    const maxReached = (property.getNumber(sh.maxCount) || Number.POSITIVE_INFINITY) <= objects.length

    return {
      ...currentProperty,
      objects,
      maxReached,
    }
  })

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: {
        ...focusNodeState,
        properties,
      },
    },
  }
}

export function editorLoading(state: FormState, { editor }: { editor: NamedNode }): FormState {
  return {
    ...state,
    editors: {
      ...state.editors,
      [editor.value]: {
        loaded: false,
      },
    },
  }
}

export function editorLoaded(state: FormState, { editor }: { editor: NamedNode }): FormState {
  return {
    ...state,
    editors: {
      ...state.editors,
      [editor.value]: {
        loaded: true,
      },
    },
  }
}

export function resetEditors(state: FormState): FormState {
  return {
    ...state,
    editors: {},
  }
}

export function initialize(state: FormState, params: { focusNode: FocusNode; shape: Shape }): FormState {
  const { focusNode, shape } = params

  if (shape.targetClass) {
    focusNode.addOut(ns.rdf.type, shape.targetClass.id)
  }

  return {
    ...state,
    focusNodes: {
      ...state.focusNodes,
      [focusNode.value]: initialiseFocusNode({
        shape,
        editors: [...state.editorMap.values()],
        compoundEditors: [...state.compoundEditorMap.values()],
        focusNode,
      }),
    },
  }
}
