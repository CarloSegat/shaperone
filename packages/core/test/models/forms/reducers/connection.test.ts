import { describe, it } from 'mocha'
import { connect, disconnect } from '../../../../models/forms/reducers/connection'
import { expect } from 'chai'

describe('core/models/forms/reducers/connection', () => {
  describe('connect', () => {
    it('creates a state for given form', function () {
      // given
      const before = {
        instances: new Map(),
        valueEditors: [],
      }
      const form = {}

      // when
      const after = connect(before, form)

      // then
      expect(after.instances.has(form))
      expect(after.instances.get(form)).to.matchSnapshot(this)
    })
  })

  describe('disconnect', () => {
    it('removes own state', () => {
      // given
      const form = {}
      const before = {
        instances: new Map(),
        valueEditors: [],
      }
      before.instances.set(form, {})

      // when
      const after = disconnect(before, form)

      // then
      expect(after.instances.size).to.eq(0)
    })
  })
})
