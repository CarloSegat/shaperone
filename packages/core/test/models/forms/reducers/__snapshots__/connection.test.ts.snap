exports["core/models/forms/reducers/connection connect creates a state for given form"] = {
	"shapes": [],
	"focusNodes": {},
	"focusStack": []
};

